File: codegen/template/go/lexer.go
package _go

import (
	"errors"
	"unicode"
)

// Lexer
type Lexer struct {
	Chars     CharIterator
	automatas map[string]Automata
}

func NewLexer(chars CharIterator) *Lexer {
	return &Lexer{
		Chars: chars,
		automatas: map[string]Automata{
			"DOG": &DOG{},
		},
	}
}

func (l *Lexer) NextToken() (t *Token, err error) {
	if Peek() == 0 {
		return nil, errors.New("reach EOF")
	}
	l.skipWhitespace()
	idx := CurrentIndex()

	for k, a := range l.automatas {
		err = RunGreedy(l.Chars)
		if err != nil {
			SetIndex(idx)
		} else {
			var tv = SubString(idx, CurrentIndex())
			idx = CurrentIndex()

			t = &Token{TokenType(k), TokenValue(tv)}
			err = nil
			break
		}
	}

	l.skipWhitespace()

	return
}

func (l *Lexer) HasNext() bool {
	return Peek() != 0
}

func (l *Lexer) skipWhitespace() {
	c := Peek()
	for unicode.IsSpace(c) {
		NextChar()
		c = Peek()
	}
}

// Token
type TokenType string
type TokenValue string
type Token struct {
	Type  TokenType
	Token TokenValue
}

File: codegen/template/cpp/Lexer.cpp
//
// Created by zhong on 2019/9/23.
//

#include "Lexer.h"

#include <utility>

Lexer::Lexer(std::shared_ptr<CharReader> charReader) : charReader(std::move(charReader)) {
    this->automatas.push_back(std::make_unique<DOG>());

}

Token Lexer::nextToken() {
    if (!this->hasNext()) {
        throw std::runtime_error("reach EOF");
    }
    this->skipWhitespace();

    for (const auto &a : this->automatas) {
        auto o = this->charReader->curIdx();
        a->runGreedy(this->charReader);
        auto n = this->charReader->curIdx();

        if (o == n) {
            continue;
        } else {
            auto res = Token(a->name(), CharReader::subString(o, n));
            this->skipWhitespace();
            return res;
        }
    }

    throw std::runtime_error("unexpected token");
}

bool Lexer::hasNext() {
    return this->charReader->peek() != 0;
}

void Lexer::skipWhitespace() {
    auto p = this->charReader->peek();
    while (std::iswspace(p)) {
        this->charReader->nextChar();
        p = this->charReader->peek();
    }
}


Token::Token(TokenType type, TokenValue token) : type(std::move(type)), token(std::move(token)) {}

